# Home assignment 02: Laplace distribution

Today your goal is to build a class for Laplace distribution. The part of the notebook copies the one from the practice session.

## Loading data

In [1]:
import numpy as np
import pandas as pd
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import datasets

In [2]:
matplotlib.rcParams['font.size'] = 11

First to load dataset we're going to use [`sklearn`](https://scikit-learn.org/stable/) package which we will extensively use during the whole course.

`sklearn` implement most of classical and frequently used algorithms in Machine Learning. Also it provides [User Guide](https://scikit-learn.org/stable/user_guide.html) describing principles of every bunch of algorithms implemented.

As an entry point to main `sklearn`'s concepts we recommend [getting started tutorial](https://scikit-learn.org/stable/getting_started.html) (check it out yourself). [Further tutorials](https://scikit-learn.org/stable/tutorial/index.html) can also be handy to develop your skills.

In [3]:
dataset = datasets.load_iris()

print(dataset.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

If you aren't familiar with Iris dataset - take a minute to read description above (as always [more info about it in Wikipedia](https://en.wikipedia.org/wiki/Iris_flower_data_set)).

__TL;DR__ 150 objects equally distributed over 3 classes each described with 4 continuous features

Just pretty table to look at:

In [4]:
ext_target = dataset.target[:, None]
pd.DataFrame(
    np.concatenate((dataset.data, ext_target, dataset.target_names[ext_target]), axis=1),
    columns=dataset.feature_names + ['target label', 'target name'],
)


sepal length (cm) sepal width (cm) petal length (cm) petal width (cm)  \
0                 5.1              3.5               1.4              0.2   
1                 4.9              3.0               1.4              0.2   
2                 4.7              3.2               1.3              0.2   
3                 4.6              3.1               1.5              0.2   
4                 5.0              3.6               1.4              0.2   
..                ...              ...               ...              ...   
145               6.7              3.0               5.2              2.3   
146               6.3              2.5               5.0              1.9   
147               6.5              3.0               5.2              2.0   
148               6.2              3.4               5.4              2.3   
149               5.9              3.0               5.1              1.8   

    target label target name  
0              0      setosa  
1              0      setosa  
2              0      setosa  
3              0      setosa  
4              0      setosa  
..           ...         ...  
145            2   virginica  
146            2   virginica  
147            2   virginica  
148            2   virginica  
149            2   virginica  

[150 rows x 6 columns]

Now give distinct names to the data we will use

In [5]:
features = dataset.data
target = dataset.target

features.shape, target.shape
features

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

__Please, remember!!!__

Anywhere in our course we have an agreement to shape design matrix (named `features` in code above) as

`(#number_of_items, #number_of_features)` if not stated explicitly.

## Distribution implementation

Let's implement class taking list of feature values, estimating Laplace distribution params and able to give probability density of any given feature value.

The file downloaded below contains the template for your class.

In [6]:
!wget https://raw.githubusercontent.com/girafe-ai/ml-course/23s_dd_ml/homeworks/hw02_laplace/distribution.py

--2025-01-25 22:11:51--  https://raw.githubusercontent.com/girafe-ai/ml-course/23s_dd_ml/homeworks/hw02_laplace/distribution.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1477 (1.4K) [text/plain]
Saving to: ‘distribution.py’

distribution.py     100%[===================>]   1.44K  --.-KB/s    in 0s      

2025-01-25 22:11:52 (20.7 MB/s) - ‘distribution.py’ saved [1477/1477]



Denote the Laplace distribution $\mathcal{L}(\mu, b)$ PDF, where $\mu$ stand for location (loc), and $b$ stands for scale:
$$
f(x|\mu, b) = \frac{1}{2b}\exp(-\frac{|x - \mu|}{b})
$$
Let's implement the `LaplaceDistribution` class. (Of course in practice one could always use something like `scipy.stats.laplace`).

Please note, that making computations with log probabilities is more stable.


#### Description [from Wikipedia](https://en.wikipedia.org/wiki/Laplace_distribution#Statistical_inference):

Given $n$ independent and identically distributed samples $x_1, x_2, ..., x_n$, the maximum likelihood (MLE) estimator of $\mu$ is the sample median:
$$
\hat{\mu} = \mathrm{median}(x).
$$



The MLE estimator $b$ is the mean absolute deviation from the median
$$
\hat{b} = \frac{1}{n} \sum_{i = 1}^{n} |x_i - \hat{\mu}|.$$

revealing a link between the Laplace distribution and least absolute deviations.
A correction for small samples can be applied as follows:
$\hat{b}^* = \hat{b} \cdot n/(n-2)$


In [7]:
# Run some setup code for this notebook.
import random
import numpy as np
import matplotlib.pyplot as plt

# Some more magic so that the notebook will reload external python modules;
# see http://stackoverflow.com/questions/1907993/autoreload-of-modules-in-ipython
%load_ext autoreload
%autoreload 2

In [8]:
# This dirty hack might help if the autoreload has failed for some reason
try:
    del LaplaceDistribution
except:
    pass

from distribution import LaplaceDistribution

SyntaxError: invalid syntax (distribution.py, line 23)

In [207]:
#Please note, that making computations with log probabilities is more stable

import numpy as np

class LaplaceDistribution:
    @staticmethod
    def mean_abs_deviation_from_median(x: np.ndarray):
        '''
        Args:
        - x: A numpy array of shape (n_objects, n_features) containing the data
          consisting of num_train samples each of dimension D.
        '''
        ####
        # Do not change the class outside of this block
        # Your code here
        b=np.zeros(len(x))
        temp=0
        if x.ndim==1:
          med=np.median(x)
          for feat in range(len(x)):
            temp+=abs(x[feat]-med)
          b=temp/(len(x))
        else:
          for obj in range(len(x)):
            temp=0
            med=np.median(x[obj])
            for feat in range(len(x[obj])):
              temp+=abs(x[obj][feat]-med)
            b[obj]=temp/(len(x[obj]))
        return b
        ####

    def __init__(self, features):
        '''
        Args:
            feature: A numpy array of shape (n_objects, n_features). Every column represents all available values for the selected feature.
        '''
        ####
        # Do not change the class outside of this block
        if features.ndim==1:
          features=np.array([features])
        else:
          features=features.T
        med=np.zeros(len(features))
        for obj in range(len(features)):
          med[obj]=np.median(features[obj])
        self.loc = med # YOUR CODE HERE
        self.scale = self.mean_abs_deviation_from_median(features)# YOUR CODE HERE

        ####


    def logpdf(self, values):
        '''
        Returns logarithm of probability density at every input value.
        Args:
            values: A numpy array of shape (n_objects, n_features). Every column represents all available values for the selected feature.
        '''
        ####
        # Do not change the class outside of this block
        ans=[]
        values=values.T
        for i in range(len(self.scale)):
          ans.append((np.log(1/(2*self.scale[i])*np.exp(-np.divide(abs(values[i].T-self.loc[i]),self.scale[i])))))
        return list(map(lambda x: list(x),list(zip(ans[0],ans[1]))))
        ####


    def pdf(self, values):
        ####
        # Do not change the class outside of this block
        ans=[]
        values=values.T
        for i in range(len(self.scale)):
          ans.append(((1/(2*self.scale[i])*np.exp(-np.divide(abs(values[i].T-self.loc[i]),self.scale[i])))))
        return list(map(lambda x: list(x),list(zip(ans[0],ans[1]))))
        ####


In [205]:
my_distr_2 = LaplaceDistribution(features[:, :2])
print(my_distr_2.logpdf(features[:5, :2]))

[[-0.1105657168116308, -0.583709268125844], [1.139434283188365, -0.08370926812584586], [-0.11056571681163638, 0.9162907318741551], [-0.7355657168116397, 0.4162907318741547], [0.514434283188367, -1.0837092681258444]]


### Distribution parameters check

In [208]:
import scipy

loc0, scale0 = scipy.stats.laplace.fit(features[:, 0])
loc1, scale1 = scipy.stats.laplace.fit(features[:, 1])

# 1d case
my_distr_1 = LaplaceDistribution(features[:, 0])

# check the 1d median (loc parameter)
assert np.allclose(my_distr_1.loc, loc0), '1d distribution median error'
# check the 1d scale (loc parameter)
assert np.allclose(my_distr_1.scale, scale0), '1d distribution scale error'


# 2d case
my_distr_2 = LaplaceDistribution(features[:, :2])

# check the 2d median (loc parameter)
assert np.allclose(my_distr_2.loc, np.array([loc0, loc1])), '2d distribution median error'
# check the 2d median (loc parameter)
assert np.allclose(my_distr_2.scale, np.array([scale0, scale1])), '2d distribution scale error'



print('Seems fine!')

Seems fine!


In [141]:
print(my_distr_2.scale, np.array([scale0, scale1]))

[0.16 0.2 ] [0.16 0.2 ]


### Distribution logpdf check

In [209]:
_test = scipy.stats.laplace(loc=[loc0, loc1], scale=[scale0, scale1])

assert np.allclose(
    my_distr_2.logpdf(features[:5, :2]),
    _test.logpdf(features[:5, :2])
), 'Logpdfs do not match scipy results!'
print('Seems fine!')

Seems fine!


In [165]:
print(_test.logpdf(features[:5, :2]))
print('\n\n\n\n\n')
ans=my_distr_2.logpdf(features[:5, :2])
print(list(map(lambda x: np.array(x),list(zip(ans[0],ans[1])))))

[[-0.11056572 -0.58370927]
 [ 1.13943428 -0.08370927]
 [-0.11056572  0.91629073]
 [-0.73556572  0.41629073]
 [ 0.51443428 -1.08370927]]






[array([-0.11056572, -7.61056572]), array([-7.58370927, -0.08370927])]


In [164]:
i=0
features=features[:5, :2]
1/(2*my_distr_2.scale[i])*np.exp(-np.divide(abs(features[i]-my_distr_2.loc[i]),my_distr_2.scale[i]))
my_distr_2.scale[i]

0.15999999999999998

Congratulations! Please, paste the `LaplaceDistribution` class code into the py file and submit it to the contest system.